File: src/main/java/com/beust/klaxon/Json.java
package com.beust.klaxon;

/**
 * @author Cedric Beust <cedric@beust.com>
 * @since 12/07/2017
 */

import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Target(FIELD)
@Retention(RUNTIME)
public @interface Json {
    String name() default "";
}

File: src/main/kotlin/com/beust/klaxon/Json.kt
package com.beust.klaxon

import java.lang.IllegalArgumentException
import java.math.BigInteger

fun convert(value: Any?) : Any? = when (value) {
        is Int -> value
        is Long -> value
        is String -> value
        is Boolean -> value
        is Float -> value
        is Double -> value
        is BigInteger -> value
        is JsonObject -> value
        is JsonArray<*> -> value
        null -> null
        else -> throw IllegalArgumentException("Unrecognized type: " + value)
    }

class JSON() {

    fun array(vararg args: Any?) : JsonArray<Any?> = JsonArray(args.map(::convert))

    fun array(args: List<Any?>) : JsonArray<Any?> = JsonArray(args.map(::convert))

    // we need this as now JsonArray<T> is List<T>
    fun <T> array(subArray : JsonArray<T>) : JsonArray<JsonArray<T>> = JsonArray(listOf(subArray))

    fun obj(vararg args: Pair<String, *>): JsonObject =
            JsonObject(linkedMapOf(*args).mapValues {convert(it.value)})
}

/**
 * Main entry point.
 */
fun <T> json(init : JSON.() -> T) : T {
    return JSON().init()
}
